docs/runtime-model.md

```diff
@@ -0,0 +1,356 @@
+
+Tasks are submitted to an executor using `tokio::spawn` or by calling a `spawn`
+method on an executor object. The `poll` function drives the task. No work is
+done without calling `poll`. It is the executor's job to call `poll` on the task
+until `Ready(())` is returned.
+
+`MyTask` will receive a value from `my_resource` and process it. Once the value
+has been processed, the task has completed its logic and is done. This is
+represented by returning `Ok(Async::Ready(()))`.
+
+However, in order to complete processing, the task depends on `my_resource`
+providing a value. Given that `my_resource` is a non-blocking task, it may or
+may not be ready to provide the value when `my_resource.poll()` is called. If it
+is ready, it returns `Ok(Async::Ready(value))`. If it is not ready, it returns
+`Ok(Async::NotReady)`.
+
+When the resource is not ready to provide a value, this implies that the task
+itself is not ready to complete and the task's `poll` function returns
+`NotReady` as well.
+
+At some point in the future, the resource will become ready to provide the
+value. The resource uses the task system to signal to the executor that it is
+ready. The executor schedules the task, which leads to `MyTask::poll` being
+called again. This time, given that `my_resource` is ready, the value will be
+returned from `my_resource.poll()` and the task is able to complete.
+
+[Future]: https://docs.rs/futures/0.1/futures/future/trait.Future.html
+
+#### Cooperative scheduling
+
+Cooperative scheduling is used to schedule tasks on executors. A single executor
+is expected to manage many tasks across a small set of threads. There will be
+a far greater number of tasks then threads. There also is no pre-emption. This
+means that when a task is scheduled to execute, it blocks the current thread
+until the `poll` function returns.
+
+Because of this, it is important for implementations of `poll` to only execute
+for very short periods of time. For I/O bound applications, this usually happens
+automatically. However, if a task must run a longer computation, it should defer
+work to a [blocking pool] or breakup the computation into smaller chunks and
+[yield] back to the executor after each chunk.
+
+[blocking pool]: #
+[yield]: #
+
+### Task system
+
+The task system is the system by which resources notify executors of readiness
+changes. A task is composed of non-blocking logic that consume resources. In the
+example above, `MyTask` uses a single resource, `my_resource`, but there is no
+limit to the number of resources that a task can consume.
+
+When a task is executing and attemepts to use a resource that is not ready, it
+becomes *logically* blocked on that resource, i.e., the task is not able to make
+further progress until that resource becomes ready. Conversely, if a task does
+not attempt to consume a resource, it is not blocked on that resource, i.e.,
+thee readiness of that resource does not impact the ability of the task to make
+forward progress.
```

Comment:
I had to read this paragraph a couple of times to grok it. I think it might be clearer to remove the  sentence starting with "Conversely" and combine the first sentence here with the next paragraph.